src/primary_names.lua

```diff
@@ -0,0 +1,301 @@
+local arns = require("arns")
+local balances = require("balances")
+local utils = require("utils")
+local primaryNames = {}
+
+local PRIMARY_NAME_COST = 100000000 -- 100 IO
+
+---@alias WalletAddress string
+---@alias ArNSName string
+
+---@class PrimaryNames
+---@field owners table<WalletAddress, PrimaryName> - map indexed by owner address containing the primary name and all metadata, used for reverse lookups
+---@field names table<ArNSName, WalletAddress> - map indexed by primary name containing the owner address, used for reverse lookups
+---@field requests table<WalletAddress, PrimaryNameRequest> - map indexed by owner address containing the request, used for pruning expired requests
+
+PrimaryNames = PrimaryNames or {
+	requests = {},
+	names = {},
+	owners = {},
+}
+
+---@class PrimaryName
+---@field name ArNSName
+---@field baseName ArNSName
+---@field startTimestamp number
+
+---@class PrimaryNameWithOwner
+---@field name ArNSName
+---@field baseName ArNSName
+---@field owner WalletAddress
+---@field startTimestamp number
+
+---@class PrimaryNameRequest
+---@field name ArNSName -- the name being requested
+---@field baseName ArNSName -- the base name, identified when creating the name request
+---@field initiator WalletAddress -- the process id that made the request
+---@field startTimestamp number -- the timestamp of the request
+---@field endTimestamp number -- the timestamp of the request expiration
+
+--- Creates a transient request for a primary name. This is done by a user and must be approved by the name owner of the base name.
+--- @param name string -- the name being requested, this could be an undername provided by the ant
+--- @param initiator string -- the address that is creating the primary name request, e.g. the ANT process id
+--- @param timestamp number -- the timestamp of the request
+--- @return PrimaryNameRequest|PrimaryNameWithOwner primaryNameRequestOrWithOwner - the request created, or the primary name with owner data if the request is approved
```

Comment:
Nit: The class for PrimaryNameRequest doesn't exactly map to the structure that's conveyed in the final return - it seems you've unpacked the request into the type definition when the request itself is actually returned in the "request" field.